### DISCUSSION TITLE
When should Operations return a Collection?

### DISCUSSION THREAD

**Discussion by AlexandruGG** [loophp/collection#107] [General]
<img width="848" alt="Screenshot 2021-06-13 at 21 39 48" src="https://user-images.githubusercontent.com/34811569/121821273-efc53300-cc8f-11eb-9b99-966ef6b77d0a.png">

---
Based on the above I thought I'd start a wider discussion, with the main question: **when should operations return a collection object**?

Here's what I think: I appreciate the consistency in knowing that a lot of the operations return the same object. There is definitely some benefit to that. However, I think even more important elements can be:

- _convenience_ -> what requires less effort on the user side when using the Collection library?
- _familiarity_ -> what will the user be familiar with from other packages/programming languages?
- _intuition_ -> when a user reads a method's name and looks at an example, does it come natural what the method is doing?
- _use cases_ -> how is the operation going to be used in conjunction with other methods?

Based on the above, I found a nice little principle that I feel fits quite well. Let me know if it makes sense: _operations should return a collection object IF they are meant to be chained_. With this I present a couple examples:

```PHP
// Classic usage, makes complete sense for these methods to return the same object to allow chainability
Collection::fromIterable([1, 2, 3])
    ->map(...)
    ->flatten(...)
    ->filter(...)
;


// In this usage, it's quite hard to find a use case where you would want to continue to chain anything after `contains`
Collection::fromIterable([1, 2, 3])
    ->contains(2) // [true] -> we have a bool so far
    ->whatCanIChainHere()? // -> the only thing I might want to do here is call `current` to get the value
;
```

Let me know if these things make sense. I can see a few methods which would benefit from having their return type changed, like `contains`, `has`, `nullsy`, `falsy`, `truthy`. As a user I think I would expect a simple boolean from all of these. 

Others which are a bit more debatable would be `first`, `get`, `head`, where in many cases a user would expect to get the value. Though for these we could argue that you might still want to chain something.

**Comment by AlexandruGG** [comment]
@drupol let me know if this makes sense when you have a chance to go over it and think about it

**Reply by drupol** [reply]
Sure, let me think about it carefully :)

**Reply by AlexandruGG** [reply]
@drupol hey, did you have a chance to go through this? 😄

**Reply by drupol** [reply]
Not yet but I will for sure!

**Comment by drupol** [comment]
We should categorize 3 types of return:

1. A scalar value; `int` or `bool` for example: `::count()`, `::has()`, ...
2. A plain collection
3. A collection of collection; eg: `::span()`, `::partition()`

Do you see something else to add?

**Reply by AlexandruGG** [reply]
I like this approach

**Reply by AlexandruGG** [reply]
@drupol are you ok with us making the change for the scalars for the next release (i.e. 4.2)?

I can prepare a PR for that if you give me the green light

**Reply by drupol** [reply]
Sure! you can prepare any PRs, submit them as draft, and then we discuss them. Just like I did for #119 and #116.

**Comment by drupol** [comment]
Should `::first()` and `::last()` do the same?

**Reply by AlexandruGG** [reply]
sorry I think I missed this.

`first`, `get`, `head`, `last` could also be decent candidates for this but I'm not too sure, I think for now let's keep them as is because it allows a bit more flexibility. For example I might want to take the first item, apply some transformation, and only then extract it via `current`

**Reply by drupol** [reply]
Fair enough.
